prac_07/language_file_reader.py

```diff
@@ -0,0 +1,89 @@
+"""
+CP1404/CP5632 Practical
+File and class example - opens/reads a file, stores in objects of custom class
+(contains multiple versions for demonstration: using csv and namedtuple)
+"""
+
+import csv
+from collections import namedtuple
+
+from programminglanguage import ProgrammingLanguage
+
+
+def main():
+    """Read file of programming language details, save as objects, display."""
+    languages = []
+    # Open the file for reading
+    in_file = open('languages.csv', 'r')
+    # File format is like: Language,Typing,Reflection,Year
+    # 'Consume' the first line (header) - we don't need its contents
+    in_file.readline()
+    # All other lines are language data
+    for line in in_file:
+        # print(repr(line))  # debugging
+        # Strip newline from end and split it into parts (CSV)
+        parts = line.strip().split(',')
+        # print(parts)  # debugging
+        # Reflection is stored as a string (Yes/No) and we want a Boolean
+        reflection = parts[2] == "Yes"
+        pointer_arithmetic = parts[4] == "Yes"
+        # Construct a ProgrammingLanguage object using the elements
+        # year should be an int
+        language = ProgrammingLanguage(parts[0], parts[1], reflection, int(parts[3]), pointer_arithmetic)
+        # Add the language we've just constructed to the list
+        languages.append(language)
+    # Close the file as soon as we've finished reading it
+    in_file.close()
+
+    # Loop through and display all languages (using their str method)
+    for language in languages:
+        print(language)
+
+
+main()
+
+
+def using_csv():
+    """Language file reader version using the csv module."""
+    # First, open the file for reading - note: specify newline
+    # to avoid quoted \n in strings being considered a new record
+    in_file = open('languages.csv', 'r', newline='')
+    in_file.readline()
+    reader = csv.reader(in_file)  # use default dialect, Excel
+    for row in reader:
+        print(row)
+    in_file.close()
+
+
+# using_csv()
+
+
+def using_namedtuple():
+    """Language file reader version using a named tuple."""
+    in_file = open('languages.csv', 'r', newline='')
+    file_field_names = in_file.readline().strip().split(',')
+    print(file_field_names)
+    # Language will be a new subclass of the tuple data type class
+    Language = namedtuple('Language', 'name, typing, reflection, year')
```

Comment:
Could be wrong, but I think you are also supposed to make pointer_arithmetic work for these other functions. Its a quick alteration

Reply:
Definitely overlooked given the rest of the tasks this week, updated now!